tests/encryption_test.go

```diff
@@ -0,0 +1,239 @@
+
+	for _, commandArgs := range prepKeyStoresCommandChain {
+		_, err := utils.RunAndLogCommandOutput(exec.Command(commandName, commandArgs...))
+		if err != nil {
+			suite.FailNow("Error while preparing the key-stores", err)
+		}
+	}
+
+	suite.Assert().FileExists(suite.keyStorePath)
+	suite.Assert().FileExists(suite.trustStorePath)
+	suite.Assert().FileExists(certPath)
+}
+
+// launches a Spark application based on `sparkConf` and asserts its successful completion
+func assertSparkApp(suite *SparkEncryptionSuite, sparkConf map[string]string, args []string) {
+	counter++
+	appName := fmt.Sprintf("spark-mock-task-runner-encrypted-%d", counter)
+
+	sparkApp := utils.SparkJob{
+		Name:     appName,
+		Template: "spark-mock-task-runner-encrypted.yaml",
+		Params: map[string]interface{}{
+			"Args":         args,
+			"SparkConf":    sparkConf,
+			"SparkSecrets": suite.sparkSecrets,
+			"SslSecrets":   suite.sslSecrets,
+		},
+	}
+
+	suite.Assert().NoError(suite.operator.SubmitJob(&sparkApp))
+	defer suite.operator.DeleteJob(sparkApp)
+
+	// when ssl is configured, check Spark UI is accessible via https on 4440 port
+	if _, present := sparkConf["spark.ssl.enabled"]; present {
+		checkSparkUI(appName, sparkApp, suite)
+	}
+
+	suite.Assert().NoError(suite.operator.WaitUntilSucceeded(sparkApp))
+}
+
+func checkSparkUI(appName string, sparkApp utils.SparkJob, suite *SparkEncryptionSuite) {
+	if err := suite.operator.WaitForJobState(sparkApp, v1beta2.RunningState); err != nil {
+		suite.Fail("SparkApplication \"%s\" is not running", appName, err)
+	}
+	if err := utils.RetryWithTimeout(20*time.Second, 2*time.Second, func() error {
+		response, err := utils.Kubectl("exec", utils.DriverPodName(appName), "-n", sparkApp.Namespace,
+			"--",
+			"curl",
+			"--insecure", // allow insecure SSL
+			"--location", // follow redirects
+			"--include",  //include headers
+			"https://localhost:4440")
+		if err != nil {
+			return err
+		}
+		if !strings.Contains(response, "HTTP/1.1 200") {
+			return fmt.Errorf("received status code is not successful")
+		}
+		suite.Assert().Contains(response, "<title>MockTaskRunner - Spark Jobs</title>")
+		return nil
+	}); err != nil {
+		suite.Fail("Unable to access Spark UI", err)
+	}
+}
+
+// returns modified Spark configuration containing 'JAVA_TOOL_OPTIONS' set for executor env,
+// which allows to pass parameters to executor processes in order to setup RPC auth and encryption.
+// under the normal circumstances, all the properties should be copied from a driver configuration,
+// but for some reason it is not working when running on k8s, so this method can be considered as a temporary workaround.
+func copyConfToExecutorEnv(sparkConf map[string]string) map[string]string {
+	var javaOpts []string
+	for key, value := range sparkConf {
+		javaOpts = append(javaOpts, fmt.Sprintf("-D%s=%s", key, value))
+	}
+	sparkConf["spark.executorEnv.JAVA_TOOL_OPTIONS"] = strings.Join(javaOpts, " ")
```

Comment:
IIUC, this step must be performed by users in order to provide secrets to executors. Can we instead handle this situation in `entrypoint.sh` and set `--conf spark.kubernetes.executor.secretKeyRef.ENV_NAME=name:key` when driver is launched? Accordingly, we can add environment variables handling in `entripoint.sh` of the executor in a similar way we do for the Driver.

WDYT?

Reply:
This is already handled here:
https://github.com/mesosphere/kudo-spark-operator/blob/7b9d274508e93e8d78d467b10f1dc698a33d2e12/tests/templates/spark-mock-task-runner-encrypted.yaml#L63
`JAVA_TOOL_OPTIONS` here is used to pass Spark configuration parameters (`spark.authenticate`, `spark.network.crypto.enabled` etc) to the executors,  which I believe should be [copied from the driver](https://github.com/apache/spark/blob/v2.4.4/core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala#L188-L215), but for some reason, they don't.

Reply:
Ok, I think I got it: executor fails to fetch configuration because it can't authenticate with a driver itself. Logs:
```
20/02/13 09:16:38 DEBUG UserGroupInformation: PrivilegedActionException as:root (auth:SIMPLE) cause:org.apache.spark.SparkException: Exception thrown in awaitResult: 
Exception in thread "main" java.lang.reflect.UndeclaredThrowableException
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1911)
	at org.apache.spark.deploy.SparkHadoopUtil.runAsSparkUser(SparkHadoopUtil.scala:64)
	at org.apache.spark.executor.CoarseGrainedExecutorBackend$.run(CoarseGrainedExecutorBackend.scala:188)
	at org.apache.spark.executor.CoarseGrainedExecutorBackend$.main(CoarseGrainedExecutorBackend.scala:281)
	at org.apache.spark.executor.CoarseGrainedExecutorBackend.main(CoarseGrainedExecutorBackend.scala)
Caused by: org.apache.spark.SparkException: Exception thrown in awaitResult: 
	at org.apache.spark.util.ThreadUtils$.awaitResult(ThreadUtils.scala:226)
	at org.apache.spark.rpc.RpcTimeout.awaitResult(RpcTimeout.scala:75)
	at org.apache.spark.rpc.RpcEnv.setupEndpointRefByURI(RpcEnv.scala:101)
	at org.apache.spark.executor.CoarseGrainedExecutorBackend$$anonfun$run$1.apply$mcV$sp(CoarseGrainedExecutorBackend.scala:201)
	at org.apache.spark.deploy.SparkHadoopUtil$$anon$2.run(SparkHadoopUtil.scala:65)
	at org.apache.spark.deploy.SparkHadoopUtil$$anon$2.run(SparkHadoopUtil.scala:64)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1893)
	... 4 more
Caused by: java.lang.RuntimeException: java.lang.IllegalStateException: Expected SaslMessage, received something else (maybe your client does not have SASL enabled?)
```



So it is sufficient to only pass `spark.authenticate=true` together with `_SPARK_AUTH_SECRET` environment variable to the executor. When those settings are set, the executor will be able to authenticate with the driver and fetch the rest of the configuration. 
So basically, we can replace `JAVA_TOOL_OPTIONS` with a single env variable in the executor spec, e.g;
```yaml
spec:
  executor:
    env:
      - name: SPARK_AUTHENTICATE
        value: true
...
We also need to modify `entrypoint.sh` to pass the required property to the executor via `SPARK_EXECUTOR_JAVA_OPTS` based on this variable.

Reply:
OR we can just check if `_SPARK_AUTH_SECRET` is set and enable `spark.authenticate` on executors based on that variable. I would go with this approach since it will require less configuration in the application spec. WDYT?